### ISSUE TITLE
Does this use ExLlamav2's Dynamic generator (paged attention)

### ISSUE THREAD

**Issue by maharshi-openfabric** [sgsdxzy/YuE-exllamav2#17]
From what I see in the code, for Stage 1 Pipeline we have:

```python
# Forward prompt
logits = self.model.forward(incremental_ids[:, :], cache=cache, input_mask=input_mask, position_offsets=position_offsets, last_id_only=True)

# Generate until EOS or max_new_tokens
for new_tokens in tqdm(range(max_new_tokens)):

    # Transformers-equiv. CFG
    if cfg:
        cfg_scale = sample_settings.guidance_scale_seg0 if i == 0 else sample_settings.guidance_scale
        logits = logits.float()
        logits = F.log_softmax(logits, dim=-1)
        logits = cfg_scale * logits[0] + (1 - cfg_scale) * logits[1]
        logits = logits.unsqueeze(0)

    # Sample
    logits = logits.float().cpu()
    sample, _, _, _, _ = ExLlamaV2Sampler.sample(logits, gen_settings, full_ids[:1], rng.random(), self.tokenizer)
    if cfg:
        sample = torch.cat((sample, sample), dim=0)

    # Accept token
    full_ids = torch.cat((full_ids, sample), dim=-1)
    seq = torch.cat((seq, sample), dim=-1)

    # Get next logits (update cache even if sample is EOA and we don't need next logits)
    if cfg:
        input_mask = full_mask[:, : full_ids.shape[-1]]
    logits = self.model.forward(sample, cache=cache, input_mask=input_mask, position_offsets=position_offsets)

    # End on EOA
    if sample[0].item() == self.mmtokenizer.eoa:
        break

# Make sure sequence ends with EOA if we reached max_new_tokens
else:
    sample = torch.tensor([[self.mmtokenizer.eoa]] * bsz, dtype=torch.long)
    seq = torch.cat((seq, sample), dim=-1)
    # Update cache with forced token
    self.model.forward(sample, cache=cache)
```

So, my question is: This doesn't seem to use ExLlamav2's dyanmic generator i.e. paged attention and it uses the model's internal forward method to get the logits and sample from them. Can we make it faster by using dynamic generator?

Thank you for making this <3

**Comment by sgsdxzy** [comment]
No, it's bs1 decoding and paged attention doesn't make sense here. In fact in [earlier commit](https://github.com/sgsdxzy/YuE-exllamav2/commit/b7ed3a5c16f206dbd65bcc375b79cc1aed91a7af) we used the dyanmic generator, and the current implementation is much faster.
